fix: Time shifts with different granularity for ECharts

michael-s-molina opened the pull request:
### SUMMARY
There's an option in the Advanced Analytics section of Explore called Time Shift that allows an user to compare the query results with the same query dislocated in time by the amount specified in the control. The way this is implemented, is to fire a query for each specified Time Shift and later join the results with the original query. To make the join work the algorithm modifies the temporal column of the time shift results to match the one in the original query. The problem is that the algorithm is not taking into account that the time granularity and time shifts can influence on the temporal column and affect the resulting timestamps. For example, if the Time Grain is Quarter and Time Shift is 28 days, the results could have different timestamps but still belong to the same quarter. Another example is when the Time Grain is Week and after applying the time shifts, the results fall in a different year which may produce different week numbers for the timestamps. The objective of this PR is to fix these problems by generating an artificial join key that takes into consideration the Time Grain when joining the results.

The following table illustrates how the join keys are calculated depending on the Time Grain:

Time Grain | Join key
-- | --
Week | year-week
Month | year-month
Quarter | year-quarter
Year | year

This join key is only used when joining the results and is discarded in the end.

This PR also improves the Python codebase by creating a `TimeGrain` class and replacing all string literals with references to the new class.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<img width="1774" alt="Screenshot 2023-05-22 at 13 24 36" src="https://github.com/apache/superset/assets/70410625/e2bc854a-2f4b-4e5c-846b-fad2d2e1d711">
<img width="1777" alt="Screenshot 2023-05-22 at 13 26 02" src="https://github.com/apache/superset/assets/70410625/9b9d385b-5a9f-4bab-9481-6a7fa8327d1a">

### TESTING INSTRUCTIONS
1 - Choose a dataset with plenty of temporal data available
2 - Create a Line Chart (ECharts)
3 - Play with different Time Grains and Time Shifts
4 - Make sure the offsets are correctly calculated

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
fix: Time shifts with different granularity

villebro commented on superset/common/query_context_processor.py:
```diff
@@ -87,12 +87,10 @@ class QueryContextProcessor:
     to retrieve the data payload for a given viz.
     """
 
+    AGGREGATED_JOIN_COLUMN = "$aggregated_join_column"
```
In general we've assumed that double underscores mean Superset internal columns. For consistency, do you think it could be possible to call this "__aggregated_join_column"?

villebro commented on superset/common/query_context_processor.py:
```diff
@@ -328,7 +325,25 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
+
+        columns = df.columns
+        time_grain = query_object.extras["time_grain_sqla"]
+        use_aggregated_join_column = any(
+            grain in time_grain for grain in ("P1W", "P1M", "P3M", "P1Y")
+        )
```
Could this be made more explicit and at the same time slightly easier to read? I think it might improve readability if we list all the time grains that are applicable, and then just check if `time_grain` is contained in that list. Something like
```python
AGGREGATED_JOIN_GRAINS = {
    "1969-12-28T00:00:00Z/P1W",
    "1969-12-29T00:00:00Z/P1W",
    "P1W/1970-01-03T00:00:00Z",
    "P1W/1970-01-04T00:00:00Z",
    "P1W",
    "P1M",
    "P3M",
    "P1Y",
}
...
use_aggregated_join_column = time_grain in AGGREGATED_JOIN_GRAINS
```

villebro reviewed (commented):
I know it's still in draft, but a few initial thoughts

zhaoyongjie commented on superset/common/query_context_processor.py:
```diff
@@ -444,10 +460,47 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
-        rv_df = pd.concat(rv_dfs, axis=1, copy=False) if time_offsets else df
-        return CachedTimeOffset(df=rv_df, queries=queries, cache_keys=cache_keys)
+        # yearly time grain
+        return row[column_index].strftime("%Y")
```
We must AVOID hard code in the core codebase. Please fill up the more description for this PR.

michael-s-molina replied:
Hi @zhaoyongjie. Hope you're doing fine! This PR is still in draft mode, that's why there's no description yet 😉

zhaoyongjie replied:
Hi @michael-s-molina, I'm great! These part codes are really important for the post processing calculation, so we should keep more eyes on the performance and maintainability!

michael-s-molina committed:
Extracts constants

michael-s-molina committed:
Replaces literals with constants

michael-s-molina committed:
Adds tests

michael-s-molina committed:
Fixes lint errors

michael-s-molina committed:
Account for no time grain

john-bodley commented on superset/common/query_context_processor.py:
```diff
@@ -74,6 +74,19 @@
 
+AGGREGATED_JOIN_COLUMN = "__aggregated_join_column"
+
+AGGREGATED_JOIN_GRAINS = {
```
@michael-s-molina how are other time grains—day, hour, minute, etc.—handled?

michael-s-molina replied:
These time grains don't require an aggregated join column because time shifts don't influence the join criteria.

michael-s-molina commented on superset/db_engine_specs/base.py:
```diff
@@ -63,6 +63,7 @@
 from typing_extensions import TypedDict
 
 from superset import security_manager, sql_parse
+from superset.constants import TimeGrain as TimeGrainConstants
```
Because there's another class defined in this file with the same name.

michael-s-molina committed:
Fixes druid time grains

zhaoyongjie commented:
@michael-s-molina 
1) IMO, the renaming/refining of TimeGrain should be put in a separate chore pull request rather than this one.
2) I can't get you result of the screen

<img width="1545" alt="image" src="https://github.com/apache/superset/assets/2016594/33489949-dab1-4d33-9c4c-20bcd03033d5">


BTW, If you could provide a dataset for testing, it would be helpful in testing this PR.

zhaoyongjie commented:
Additionally, I believe that the time grain and time offset granularity should be consistent. We might use same dataset to discuss this topic.

michael-s-molina committed:
Removes columns used for join

michael-s-molina committed:
Gets time grain from form_data

michael-s-molina commented:
> 1. IMO, the renaming/refining of TimeGrain should be put in a separate chore pull request rather than this one.

I could do that but I guess I was excited to fix the problem and ended up doing all at once. Sorry if this makes it more difficult to review but I think the important part is that I took the time to improve the code base.

> 2. I can't get result of your screenshot. (on this PR branch)

It was not working indeed with `GENERIC_CHART_AXES` enabled because the query object didn't contain a `time_grain_sqla` field. I submitted a fix to get this from the `form_data` and now it should be working.

michael-s-molina committed:
Adds a method to get the time grain

michael-s-molina commented on superset/common/query_context_processor.py:
```diff
@@ -307,6 +324,20 @@ def _get_timestamp_format(
 
         return df
 
+    def get_time_grain(
```
The joys of Superset and its N different ways to store/retrieve the same information :facepalm:

michael-s-molina commented on superset/common/query_context_processor.py:
```diff
@@ -307,6 +324,24 @@ def _get_timestamp_format(
 
         return df
 
+    @staticmethod
+    def get_time_grain(
+        query_context: QueryContext, query_object: QueryObject
+    ) -> Optional[Any]:
+        if query_context.form_data:
+            return query_context.form_data.get("time_grain_sqla")
+
+        if (
+            query_object.columns
+            and len(query_object.columns) > 0
+            and isinstance(query_object.columns[0], dict)
+        ):
+            # If the time grain is in the columns it will be the first one
+            # and it will be of AdhocColumn type
+            return query_object.columns[0].get("timeGrain")
+
+        return query_object.extras.get("time_grain_sqla")
```
The joys of Superset and its N different ways to store/retrieve the same information 🤦

michael-s-molina committed:
Adds config to set join column producerr

michael-s-molina commented on superset/common/query_context_processor.py:
```diff
@@ -444,10 +467,48 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
+
+    @staticmethod
+    def get_aggregated_join_column(
+        row: pd.Series, column_index: int, time_grain: str
```
`TimeGrain` is a `string` because we have some configurations defined as strings and changing them would constitute breaking changes and the window is already closed.

```
TIME_GRAIN_DENYLIST: list[str] = []
TIME_GRAIN_ADDONS: dict[str, str] = {}
TIME_GRAIN_ADDON_EXPRESSIONS: dict[str, dict[str, str]] = {}
```

michael-s-molina committed:
Extracts method

zhaoyongjie commented on superset/common/query_context_processor.py:
```diff
@@ -420,21 +487,19 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
                     )
 
+                # modifies temporal column using offset
                 offset_metrics_df[index] = offset_metrics_df[index] - DateOffset(
                     **normalize_time_delta(offset)
                 )
```
the index of `offset_metrics_df` seems constructing an appropriate index for the result of dataframe, if the align of the results incorrect, I think we should find root cause from here.

michael-s-molina replied:
I'm not sure if I understand what you're suggesting. Can you explain more?

villebro commented on superset/common/query_context_processor.py:
```diff
@@ -74,6 +74,27 @@
 stats_logger: BaseStatsLogger = config["STATS_LOGGER"]
 logger = logging.getLogger(__name__)
 
+# Artificial column used for joining aggregated offset results
+AGGREGATED_JOIN_COLUMN = "__aggregated_join_column"
```
nit: temporary seems more appropriate here
```suggestion
# Temporary column used for joining aggregated offset results
AGGREGATED_JOIN_COLUMN = "__aggregated_join_column"
```

villebro commented on superset/common/query_context_processor.py:
```diff
@@ -307,6 +324,40 @@ def _get_timestamp_format(
+    def get_time_grain(
+        query_context: QueryContext, query_object: QueryObject
+    ) -> Any | None:
+        if query_context.form_data:
+            return query_context.form_data.get("time_grain_sqla")
```
I think this should be avoided. The form data has been introduced fairly recently in query_context, and isn't really meant to influence rendering of the query_object (IIRC, it was originally added to relay some context about query timeouts that are defined in the chart object). Is there some reason we are primarily looking at the form data here, and not the query object?

michael-s-molina replied:
I can remove this line given that the created function deals with the different `query_object` variations that depend on `GENERIC_CHART_AXES`. I was using the `form_data` because it was the only structure that looks consistent independently of the feature flag.

villebro replied:
I understand the confusion this causes. I'll try to prepare a follow-up that moves the relevant info from form data into query context/object and removes form data all together.

villebro commented on superset/common/query_context_processor.py:
```diff
@@ -328,7 +378,25 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
-        for offset in time_offsets:
+
+        columns = df.columns
+        time_grain = self.get_time_grain(query_context, query_object) or TimeGrain.DAY
```
Why are we defaulting to a daily time grain? If it's mandatory, I'd rather return a 400 if it's missing.

villebro commented on superset/common/query_context_processor.py:
```diff
@@ -328,7 +378,25 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
+        join_column_producer = config["TIME_GRAIN_JOIN_COLUMN_PRODUCERS"].get(
+            time_grain
+        )
+        use_aggregated_join_column = (
+            time_grain in AGGREGATED_JOIN_GRAINS or join_column_producer
+        )
```
Ok, this is a total mininit and functionally they're equivalent, but for readability I'd place the check for `join_column_producer` first, as it takes precedent in downstream logic (this code is generally pretty heavy on the eyes, so any improved readability helps):
```suggestion
        use_aggregated_join_column = (
            join_column_producer  or time_grain in AGGREGATED_JOIN_GRAINS
        )
```

villebro commented on superset/db_engine_specs/base.py:
```diff
@@ -80,25 +81,25 @@ class TimeGrain(NamedTuple):
 
 
 builtin_time_grains: dict[str | None, str] = {
-    "PT1S": __("Second"),
```
The type declaration on line 82 should be updated; the key should be `TimeGrainConstants`, right?

michael-s-molina replied:
It's actually not the class but the string values defined in it

villebro replied:
Oh right - I believe it should in fact be an enum (I'll comment on the relevant line)

villebro replied:
If you make the `Enum` change above, you should be able to change the above to 
```
builtin_time_grains: dict[Optional[TimeGrainConstants], str] = {
```

michael-s-molina replied:
Check this [comment](https://github.com/apache/superset/pull/24176#discussion_r1214443954)

villebro commented on tests/unit_tests/common/test_get_aggregated_join_column.py:
```diff
@@ -0,0 +1,52 @@
+row = [Timestamp("2020-01-07")]
+
+
+def test_week_join_column():
```
I notice we don't have any tests for `TIME_GRAIN_JOIN_COLUMN_PRODUCERS` - it'd be nice to add at least one test to guard against future regressions.

villebro reviewed (commented):
A few second pass comments

michael-s-molina committed:
Addresses comments

michael-s-molina commented:
@villebro I addressed your comments.

villebro commented on tests/unit_tests/common/test_get_aggregated_join_column.py:
```diff
@@ -0,0 +1,73 @@
+)
+
+
+def join_column_producer(row: Series, column_index: int) -> str:
+    return "CUSTOM_FORMAT"
```
style nit: it would probably be best practice to define this as a `@pytest.fixture`

michael-s-molina replied:
Done

villebro reviewed (approved):
A few small optional stylistic comments

michael-s-molina committed:
Uses a fixture in the test

villebro reviewed (approved):
LGTM!

michael-s-molina committed:
Fixes types

zhaoyongjie commented:
@villebro @michael-s-molina I'll take another look this PR, give me some time. Thanks!

zhaoyongjie commented:
@michael-s-molina @villebro 

Draft a PR to refine the join key generator. please review. Thanks both!

michael-s-molina commented:
> Draft a [PR](https://github.com/apache/superset/pull/24328/files) to refine the join key generator. please review. Thanks both!

@zhaoyongjie That's not the best way to help reviewing a PR. I spent a lot of time debugging the issue, addressing comments and executing tests using Airbnb charts to get to a solution. If you think the solution can be improved, you can make suggestions, provide more context, contribute code to the PR, etc. 

Completely ignoring the PR and opening your own is not cool 😕

zhaoyongjie commented:
@michael-s-molina 
I've made the following suggestions during the code review:

1. It would be better to keep this pull request focused on the necessary changes rather than including refactorings and unrelated fixes. Let's aim to minimize the scope of the PR.
2. We should avoid hard coding values without considering the maintainability for other engineers.

michael-s-molina commented:
@zhaoyongjie I thought I have addressed both your suggestions with my latest commits. You even removed the **Request Changes** status. Then you asked time to review the PR and suddenly opened your own which was really confusing for me.
